bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/crypto/EncryptionHandler.java

```diff
@@ -0,0 +1,307 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.paradoxalarm.internal.communication.crypto;
+
+import java.util.Arrays;
+
+import org.openhab.binding.paradoxalarm.internal.util.ParadoxUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used to encrypt and decrypt communication from/to Paradox system. Singleton pattern.
+ *
+ * Paradox encryption is using Rijndael 256-key expansion alghoritm.
+ * When key is changed the updateKey(byte[]) method needs to be called in front.
+ * Encrypt and Decrypt methods use the expandedKey field to do their job.
+ *
+ * The first packet sent to Paradox is the IP150 password as bytes, extended to 32 bytes with 0xEE.
+ * The first response contains the key that will be used for the rest of communication.
+ *
+ * Most of the coding is copy from Python / rewrite in Java from second link of jpbaracca's PAI repository. Probably
+ * some of the variables can be named better but I don't understand this code in it's full scope so I preferred to keep
+ * it as it is.
+ *
+ * @author Konstantin Polihronov - Initial contribution
+ *
+ * @see <a href=https://www.samiam.org/key-schedule.html>Sam Trendholme's page about AES</a>
+ * @see <a href=https://github.com/ParadoxAlarmInterface/pai>Github of jpbaracca's work - ParadoxAlarmInterface in
+ *      python</a>
+ */
+public class EncryptionHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(EncryptionHandler.class);
+
+    private static final int KEY_ARRAY_LENGTH = 32;
+    private static final int TABLE_SIZE = 256;
+    private static final int KEY_LENGTH = 240;
+    private static final int PAYLOAD_RATE_LENGTH = 16;
+
+    private static EncryptionHandler instance;
+
+    private int[] lTable = new int[TABLE_SIZE];
+    private int[] aTable = new int[TABLE_SIZE];
+
+    private int[] expandedKey = new int[KEY_LENGTH];
+
+    private EncryptionHandler() {
+        generateTables();
+    }
+
+    public static EncryptionHandler getInstance() {
+        if (instance == null) {
+            synchronized (EncryptionHandler.class) {
+                instance = new EncryptionHandler();
+            }
+        }
+        return instance;
+    }
```

Comment:
This class is not thread safe, and as such you should not provide a singleton instance for it.

Reply:
If I make the instance initialized as per NonNullByDefault statically and do the only updater, i.e. updateKey() method synchronized can we consider the class thread safe? Based on the logic we set the key twice, once in the very beginning of the communication and second - after the initial response is received with the key provided by the system. I personally don't expect anything to change the instance at any other point in time. All the rest is manipulation on input using this key. If I'm thinking in the wrong direction, could you elaborate a bit more?

Reply:
The main thing I am worried about is that `updateKey` will get called while some encryption process is going on. It would be better if instead of making this class a singleton you can make this class readonly and do your `updateKey` logic in the constructor. Code that wants to change the encryption key would need to construct a new instance of this class.

If this class was immutable I would have no problem with you making it a singleton, the problem is that this class holds state. As a general principle singleton class should not be mutable, too many problems can arise otherwise.

Reply:
I understand the concern. I will try to make the class immutable even though it's completely safe to leave it as it is because 1. the logon sequence is always a sequence of sending one byte and receiving one response byte for this request and 2. you cannot have more than one connection to the Paradox system (it's forbidden from the other end). 
The sequence is something like this:
Update the key in the singleton instance with the password provided.
Create first packet with payload the initial key (the password). 
Encrypt the packet with initial key.
Receive the response
Decrypt the response with the initial key
Extract the new key from the first response
Update the key in the encryption handler
Continue the logon process by building, encrypting and sending the second packet from the sequence.

The only danger is if we have a lifecycle issue with bridge handler, i.e. if we have two active bridge handlers at the same time. Then the key can be modified in non-desired way but in this case many other things will go wrong anyway...

Reply:
OK I looked a bit in the code. What do you think about this approach:
Changing the private constructor to always receive the key. In the updateKey method we create a new instance with proper key, expand the key and whatever is necessary, assign it to the static singleton field and return the instance back from updateKey method.
With this approach whichever thread received the previous instance will be able to finish it's job with old state while the new jobs will run with the new instance.
I'd really like to use singleton because I don't want to create encryption instances on every packet as it may become expensive overhead to always generate the key, etc...

What do you think?

Reply:
... or the other way would be to ensure that in each method I get the reference to the field mutable field at the start and then continue with the local variable until the end of the method execution I guess... I've pushed the first version. Please let me know which one would you prefer.

Reply:
I thing what you pushed is satisfactory.